pkg/lightning/backend/tidb/tidb.go

```diff
@@ -171,123 +171,154 @@ func (b *targetInfoGetter) FetchRemoteDBModels(ctx context.Context) ([]*model.DB
-// It implements the `backend.TargetInfoGetter` interface.
-// TODO: refactor
-func (b *targetInfoGetter) FetchRemoteTableModels(ctx context.Context, schemaName string) ([]*model.TableInfo, error) {
-	var err error
-	results := []*model.TableInfo{}
+// exported for test.
+var (
+	FetchRemoteTableModelsConcurrency = 8
+	FetchRemoteTableModelsBatchSize   = 32
+)
+
+// FetchRemoteTableModels implements the `backend.TargetInfoGetter` interface.
+func (b *targetInfoGetter) FetchRemoteTableModels(
+	ctx context.Context,
+	schemaName string,
+	tableNames []string,
+) (map[string]*model.TableInfo, error) {
+	tableInfos := make([]*model.TableInfo, len(tableNames))
 	logger := log.FromContext(ctx)
 	s := common.SQLWithRetry{
 		DB:     b.db,
 		Logger: logger,
 	}
 
-	err = s.Transact(ctx, "fetch table columns", func(_ context.Context, tx *sql.Tx) error {
-		var versionStr string
-		if versionStr, err = version.FetchVersion(ctx, tx); err != nil {
-			return err
+	eg, egCtx := util.NewErrorGroupWithRecoverWithCtx(ctx)
+	eg.SetLimit(FetchRemoteTableModelsConcurrency)
+	for i := 0; i < len(tableNames); i += FetchRemoteTableModelsBatchSize {
+		start := i
+		end := i + FetchRemoteTableModelsBatchSize
+		if end > len(tableNames) {
+			end = len(tableNames)
 		}
-		serverInfo := version.ParseServerInfo(versionStr)
-
-		rows, e := tx.Query(`
-			SELECT table_name, column_name, column_type, generation_expression, extra
-			FROM information_schema.columns
-			WHERE table_schema = ?
-			ORDER BY table_name, ordinal_position;
-		`, schemaName)
-		if e != nil {
-			return e
-		}
-		defer rows.Close()
-
-		var (
-			curTableName string
-			curColOffset int
-			curTable     *model.TableInfo
-		)
-		tables := []*model.TableInfo{}
-		for rows.Next() {
-			var tableName, columnName, columnType, generationExpr, columnExtra string
-			if e := rows.Scan(&tableName, &columnName, &columnType, &generationExpr, &columnExtra); e != nil {
-				return e
-			}
-			if tableName != curTableName {
-				curTable = &model.TableInfo{
-					Name:       pmodel.NewCIStr(tableName),
-					State:      model.StatePublic,
-					PKIsHandle: true,
-				}
-				tables = append(tables, curTable)
-				curTableName = tableName
-				curColOffset = 0
-			}
-
-			// see: https://github.com/pingcap/parser/blob/3b2fb4b41d73710bc6c4e1f4e8679d8be6a4863e/types/field_type.go#L185-L191
-			var flag uint
-			if strings.HasSuffix(columnType, "unsigned") {
-				flag |= mysql.UnsignedFlag
-			}
-			if strings.Contains(columnExtra, "auto_increment") {
-				flag |= mysql.AutoIncrementFlag
-			}
+		eg.Go(func() error {
+			return s.Transact(
+				egCtx, "fetch table columns",
+				func(_ context.Context, tx *sql.Tx) error {
+					args := make([]any, 0, 1+end-start)
+					args = append(args, schemaName)
+					for _, tableName := range tableNames[start:end] {
+						args = append(args, tableName)
+					}
+					//nolint:gosec
+					rows, e := tx.Query(`
```

Comment:
```suggestion
					rows, err := tx.Query(`
```
prefer following the convention to use the name `err`. below too